agents/agents-core/src/commonMain/kotlin/ai/koog/agents/core/feature/AIAgentGraphPipeline.kt

```diff
@@ -0,0 +1,260 @@
+        interceptContext: InterceptContext<TFeature>,
+        handle: suspend TFeature.(eventContext: NodeAfterExecuteContext) -> Unit
+    ) {
+        val existingHandler = executeNodeHandlers.getOrPut(interceptContext.feature.key) { ExecuteNodeHandler() }
+
+        existingHandler.afterNodeHandler = AfterNodeHandler { eventContext: NodeAfterExecuteContext ->
+            with(interceptContext.featureImpl) { handle(eventContext) }
+        }
+    }
+
+    /**
+     * Intercepts and handles node execution errors for a given feature.
+     *
+     * @param interceptContext The context containing the feature and its implementation required for interception.
+     * @param handle A suspend function that processes the node execution error within the scope of the provided feature.
+     *
+     * Example:
+     * ```
+     * pipeline.interceptNodeExecutionError(InterceptContext) { eventContext ->
+     *     logger.error("Node ${eventContext.node.name} execution failed with error: ${eventContext.throwable}")
+     * }
+     * ```
+     */
+    public fun <TFeature : Any> interceptNodeExecutionError(
+        interceptContext: InterceptContext<TFeature>,
+        handle: suspend TFeature.(eventContext: NodeExecutionErrorContext) -> Unit
+    ) {
+        val existingHandler = executeNodeHandlers.getOrPut(interceptContext.feature.key) { ExecuteNodeHandler() }
+
+        existingHandler.nodeExecutionErrorHandler =
+            NodeExecutionErrorHandler { eventContext: NodeExecutionErrorContext ->
+                with(interceptContext.featureImpl) { handle(eventContext) }
+            }
+    }
+
+    /**
+     * Installs a feature into the pipeline with the provided configuration.
+     *
+     * This method initializes the feature with a custom configuration and registers it in the pipeline.
+     * The feature's message processors are initialized during installation.
+     *
+     * @param Config The type of the feature configuration
+     * @param Feature The type of the feature being installed
+     * @param feature The feature implementation to be installed
+     * @param configure A lambda to customize the feature configuration
+     */
+    public fun <Config : FeatureConfig, Feature : Any> install(
+        feature: AIAgentGraphFeature<Config, Feature>,
+        configure: Config.() -> Unit
+    ) {
+        val config = feature.createInitialConfig().apply { configure() }
+        feature.install(
+            config = config,
+            pipeline = this,
+        )
+
+        registeredFeatures[feature.key] = config
+    }
+
+    /**
+     * Notifies all registered handlers that an agent has started execution.
+     *
+     * @param runId The unique identifier for the agent run
+     * @param agent The agent instance for which the execution has started
+     * @param strategy The strategy being executed by the agent
+     */
+    @OptIn(InternalAgentsApi::class)
+    public suspend fun onBeforeAgentStarted(
```

Comment:
I think Agent specific events are quite common for both graph and non-graph agents.

Reply:
Unfortunately, `onBeforeAgentStarted` knows a lot about internals of agent, such as existence of graph, so it was moved to graph specific pipeline. I believe we need separate onBeforeAgentStarted in non-graph pipeline